File: swift/Sources/NES/NESConfiguration.swift
import Foundation

/// Runtime options for one Nestopia session.
public struct NESConfiguration: Sendable, Equatable {
    public var romURL: URL
    public var saveDirectory: URL?
    public var automaticallyStarts: Bool
    public var showsTouchControls: Bool

    public init(
        romURL: URL,
        saveDirectory: URL? = nil,
        automaticallyStarts: Bool = true,
        showsTouchControls: Bool = true
    ) {
        self.romURL = romURL
        self.saveDirectory = saveDirectory
        self.automaticallyStarts = automaticallyStarts
        self.showsTouchControls = showsTouchControls
    }
}

public enum NESState: Equatable, Sendable {
    case idle
    case loading
    case running
    case paused
    case stopped
    case failed(String)
}

public enum NESControllerButton: Int, CaseIterable, Sendable {
    case a = 0x01
    case b = 0x02
    case select = 0x04
    case start = 0x08
    case up = 0x10
    case down = 0x20
    case left = 0x40
    case right = 0x80
}

File: android/nes/src/main/kotlin/nestopia/NESConfiguration.kt
// Copyright (C) 2026 Dominic Opitz
// SPDX-License-Identifier: GPL-2.0-or-later

package nestopia

import android.net.Uri

data class NESConfiguration(
    val romUri: Uri,
    val automaticallyStarts: Boolean = true,
    val showsTouchControls: Boolean = true,
)

sealed interface NESState {
    data object Idle : NESState

    data object Loading : NESState

    data object Running : NESState

    data object Paused : NESState

    data object Stopped : NESState

    data class Failed(
        val message: String,
    ) : NESState
}

enum class NESButton(
    val mask: Int,
) {
    A(0x01),
    B(0x02),
    Select(0x04),
    Start(0x08),
    Up(0x10),
    Down(0x20),
    Left(0x40),
    Right(0x80),
}
